### ISSUE TITLE
Warning when "target" is too high for the node version

### ISSUE THREAD

**Issue by cspotcode** [TypeStrong/ts-node#1248]
Building off of #1202, we know the maximum "target" you can use for the given node version.

Our CLI can log a warning:

```
You are using a target option that may be incompatible with your node version.  We recommend target: es2018 or lower for node vXYZ.  To disable this warning, add "ts-node": {"disableNodeCompatibilityWarnings": true} to your tsconfig file
```

### Questions

* [ ] Should this warning be logged for programmatic usage?  For `ts-node/register`?  For `--loader ts-node/esm`?  I don't think libraries should send output to stderr, but `ts-node/register` is a weird case because it's almost exclusively used from the CLI.
* [ ] Should the `disableNodeCompatibilityWarnings` option be renamed to allow more than 2 modes?
  * `"forceNodeCompatibility": "auto" | "warn" | "off"
  * `"autoNodeCompatibility": "enabled" | "warn" | "disabled"

**Comment by cspotcode** [comment]
We discussed this on Discord: https://discord.com/channels/508357248330760243/508357707602853888/843546169752813589

### Conclusions:

* Use node's warnings API: `process.emitWarning()`
  * there is already precedent for warnings to emit to stderr from here
  * node already has APIs to intercept these warnings and flags to suppress them
* Match node's behavior and emit each warning only once
* Add `warnings: boolean` to API surface
  * `false` to disable them
  * Future work can expand to suppress or allow warnings by code
* Add `onWarning: callback` to API surface
  * if specified, warnings are sent here instead of to `process.emitWarning`
* All warnings get a code `TSNODE<4-digit number>`
  * This is familiar: node's deprecation warnings are `DEP1234`; TypeScript uses `TS1234`
  * code allows easy detection in `process.on('warning'` handlers
